staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -251,16 +254,90 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 			continue
 		}
 		if evalResult != types.True {
-			if len(rule.Message) != 0 {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if compiled.MessageExpression != nil {
+				messageExpression, newRemainingBudget, msgErr := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
+				if msgErr != nil {
+					if msgErr.errType == runtimeCostErr {
+						errs = append(errs, field.InternalError(fldPath, msgErr.err))
+						return errs, -1
+					} else {
+						klog.Warning(msgErr.err)
+						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+						remainingBudget = newRemainingBudget
+					}
+				} else {
+					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
+					remainingBudget = newRemainingBudget
+				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, fmt.Sprintf("failed rule: %s", ruleErrorString(rule))))
+				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
 			}
 		}
 	}
 	return errs, remainingBudget
 }
 
+type _errType int
+
+const (
+	runtimeCostErr _errType = iota
+	evalErr
+)
+
+type messageExpressionErr struct {
+	errType _errType
+	err     error
+}
+
+// evalMessageExpression takes a context, Program and source code representing the MessageExpression, activation,
+// and remaining, runtime cost budget, returns the evaluated message and the remaining budget, or an error if one
+// occurred during evaluation.
+func evalMessageExpression(ctx context.Context, expr celgo.Program, exprSrc string, activation interpreter.Activation, remainingBudget int64) (string, int64, *messageExpressionErr) {
+	evalResult, evalDetails, err := expr.ContextEval(ctx, activation)
+	if evalDetails == nil {
+		return "", -1, &messageExpressionErr{
+			runtimeCostErr,
+			fmt.Errorf("runtime cost could not be calculated for message expression: %q", exprSrc),
+		}
+	}
+	rtCost := evalDetails.ActualCost()
+	if rtCost == nil {
+		return "", -1, &messageExpressionErr{
+			runtimeCostErr,
+			fmt.Errorf("runtime cost could not be calculated for message expression: %q", exprSrc),
+		}
+	} else {
+		if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
+			return "", -1, &messageExpressionErr{
+				runtimeCostErr,
+				errors.New("messageExpression evaluation failed due to running out of cost budget, no further validation rules will be run"),
+			}
+		}
+	}
+	if err != nil {
+		return "", remainingBudget - int64(*rtCost), &messageExpressionErr{
+			evalErr,
+			fmt.Errorf("error evaluating message expression, falling back: %v", err.Error()),
```

Comment:
nit: keep the err message consistent
```suggestion
fmt.Errorf("message expression evaluation failed due to : %v", err.Error()),
```

Reply:
Gotcha, fixed